### DISCUSSION TITLE
Does AWS has RUST device SDK?

### DISCUSSION THREAD

**Discussion by FancyQian** [aws/aws-iot-device-sdk-cpp-v2#747] [Q&A] [answered]
I know AWS IOT device SDK has C and C++ SDK, such as https://github.com/aws/aws-iot-device-sdk-embedded-C,
Right now, more and more pepole use rust develop embedded device, does AWS IOT device SDK support rust?

**Comment by bretambrose** [comment]
Nothing official, no.

I have a personal project (https://crates.io/crates/gneiss-mqtt-aws and https://crates.io/crates/gneiss-mqtt) for an async MQTT 5 client that has support for all the current authentication methods for AWS IoT Core but there's still a lot of polish remaining to be done (311 support, non-async client, dependency stripping) before I move on to service clients, Greengrass IPC, etc...

**Reply by FancyQian** [reply]
Thanks for your reply, that is good to know that, now I know your project than other rust ask one that not maintained, I need to evaluate your project, my use cases will only has auth, pub&sub form iot aws, since you project seems will become official one(I guess), then I will have more confidence with yours, can you kindly suggest me?

**Reply by bretambrose** [reply]
To be honest, I don't think it will become official any time soon if ever.  The existing device SDKs are based on the common runtime which is a large collection of native (C) libraries that implement all the actual protocol/crypto/TLS/auth functionality.  From an Amazon-internal (can't really go into details) perspective, a device Rust SDK would probably need to follow suit (may change in the future though).

**Reply by FancyQian** [reply, marked as answer]
Tks, good answer! I am thinking use non-official device rust ask in short term for simple use cases, if we have more advanced use cases I will think using rust ffi C SDK if no official RSDK available.
